htdocs/subtotals/class/commonsubtotal.class.php

```diff
@@ -0,0 +1,595 @@
+		for ($i = $line->rang-1; $i > 0; $i--) {
+			if ($this->lines[$i-1]->special_code == SUBTOTALS_SPECIAL_CODE && $this->lines[$i-1]->qty>0) {
+				if ($this->lines[$i-1]->qty <= abs($line->qty)) {
+					return price($final_amount);
+				}
+			} else {
+				$final_amount += $this->lines[$i-1]->total_ht;
+			}
+		}
+		return price($final_amount);
+	}
+
+	/**
+	 * Return the multicurrency_total_ht of lines that are above the current line (excluded) and that are not a subtotal line
+	 * until a title line of the same level is found
+	 *
+	 * @param object	$line
+	 * @return int		$total_ht
+	 */
+	public function getSubtotalLineMulticurrencyAmount($line)
+	{
+		$final_amount = 0;
+		for ($i = $line->rang-1; $i > 0; $i--) {
+			if ($this->lines[$i-1]->special_code == SUBTOTALS_SPECIAL_CODE && $this->lines[$i-1]->qty>0) {
+				if ($this->lines[$i-1]->qty <= abs($line->qty)) {
+					return price($final_amount);
+				}
+			} else {
+				$final_amount += $this->lines[$i-1]->multicurrency_total_ht;
+			}
+		}
+		return price($final_amount);
+	}
+
+	/**
+	 * Returns a form array to add a subtotal or title line
+	 *
+	 * @param Form $form
+	 * @param Translate $langs
+	 * @param string $type 'title' or 'subtotal'
+	 * @return string $formconfirm
+	 */
+	public function getSubtotalForm($form, $langs, $type, $seller)
+	{
+		$langs->load('subtotals');
+
+		if ($type == 'subtotal') {
+			$titles = $this->getPossibleTitles();
+		}
+
+		$depth_array = $this->getPossibleLevels($langs);
+
+		$tpl = dol_buildpath('/core/tpl/subtotal_create.tpl.php');
+
+		if (empty($conf->file->strict_mode)) {
+			$res = @include $tpl;
+		} else {
+			$res = include $tpl; // for debug
+		}
+
+		return $res;
+	}
+
+	/**
+	 * Retrieve the background color associated with a specific subtotal level.
+	 *
+	 * @param int $level The level of the subtotal for which the color is requested.
+	 * @return string|null The background color in hexadecimal format or null if not set.
+	 */
+	public function getSubtotalColors($level) {
+		return getDolGlobalString('SUBTOTAL_BACK_COLOR_LEVEL_'.abs($level));
+	}
+
+	/**
+	 * Retrieve current object possible titles to choose from
+	 *
+	 * @return array The set of titles, empty if no title line set.
+	 */
+	public function getPossibleTitles() {
+		$titles = array();
+		foreach ($this->lines as $line) {
+			if ($line->special_code == SUBTOTALS_SPECIAL_CODE && $line->qty > 0) {
+				$titles[$line->desc] = $line->desc;
+			}
+			if ($line->special_code == SUBTOTALS_SPECIAL_CODE && $line->qty < 0) {
+				unset($titles[$line->desc]);
+			}
+		}
+		return $titles;
+	}
+
+	/**
+	 * Retrieve the current object possible levels (defined in admin page)
+	 *
+	 * @param Translate $langs
+	 * @return array The set of possible levels, empty if not defined correctly.
+	 */
+	public function getPossibleLevels($langs) {
+		$depth_array = array();
+		for ($i = 0; $i < getDolGlobalString('SUBTOTAL_'.strtoupper($this->element).'_MAX_DEPTH', 2); $i++) {
```

Comment:
Attention, dans une boucle `for`, mettre un appel de fonction dans la deuxième partie est mauvais pour les performances puisqu'il sera exécuté à chaque itération.